include/eos/fitting/fitting.hpp

```diff
@@ -150,57 +150,51 @@ inline Eigen::VectorXf fit_shape(Eigen::Matrix<float, 3, 4> affine_camera_matrix
 };
 
 /**
- * @brief Takes a LandmarkCollection of 2D landmarks and, using the landmark_mapper, finds the
- * corresponding 3D vertex indices and returns them, along with the coordinates of the 3D points.
- *
- * The function only returns points which the landmark mapper was able to convert, and skips all
- * points for which there is no mapping. Thus, the number of returned points might be smaller than
- * the number of input points.
- * All three output vectors have the same size and contain the points in the same order.
- * \c landmarks can be an eos::core::LandmarkCollection<cv::Vec2f> or an rcr::LandmarkCollection<cv::Vec2f>.
- *
- * Notes:
- * - Split into two functions, one which maps from 2D LMs to vtx_idx and returns a reduced vec of 2D LMs.
- *   And then the other one to go from vtx_idx to a vector<Vec4f>.
- * - Place in a potentially more appropriate header (shape-fitting?).
- * - Could move to detail namespace or forward-declare.
- * - \c landmarks has to be a collection of LMs, with size(), [] and Vec2f ::coordinates.
- * - Probably model_points would better be a Vector3f and not in homogeneous coordinates?
+ * @brief Gets the vertex index from the given landmark name using the landmark mapper and if provided the landmark definitions.
  *
- * @param[in] landmarks A LandmarkCollection of 2D landmarks.
+ * The function only returns the vertex index if the landmark mapper was able to convert the name.
+ *
+ * @param[in] landmark_name Name of the landmark, often used as identifier.
  * @param[in] landmark_mapper A mapper which maps the 2D landmark identifiers to 3D model vertex indices.
- * @param[in] morphable_model Model to get the 3D point coordinates from.
- * @return A tuple of [image_points, model_points, vertex_indices].
+ * @param[in] landmark_definitions A set of landmark definitions for the model, mapping from identifiers to vertex indices.
+ * @return An optional int with the vertex index.
  */
-template <class T>
-inline auto get_corresponding_pointset(const T& landmarks, const core::LandmarkMapper& landmark_mapper,
-                                       const morphablemodel::MorphableModel& morphable_model)
+inline std::optional<int> get_vertex_index(const std::string landmark_name, const core::LandmarkMapper& landmark_mapper, 
+                            cpp17::optional<std::unordered_map<std::string, int>> landmark_definitions)
```

Comment:
I think we probably want to pass the `landmark_definitions` as a reference too, so just add `const ... &`.

Reply:
Out of curiosity what is the benefit behind having it as a reference here?

Reply:
That's a good question and I was actually trying to google it before making my earlier comment. If `std::optional` is a "pointer type" (i.e. it holds a pointer to the underlying data, and when you pass it around, even by value, only the address/pointer is passed around), then we might as well pass it by value (without the `const ... &`), as it will then make no difference whether we pass around a pointer (32/64 bit) or a reference (32/64 bit). However if all the data is stored in the `optional` object directly, then we might potentially copy around a big/heavy object.
I *think* `optional` would be a pointer type... but I couldn't really find out. So to be safe I'd just pass it by `const&`, even if perhaps not needed, I don't think it has a drawback either.

Reply:
Thanks for the explanation, I can see where the issue is. Using `const&` makes sure we avoid passing the whole object if `optional` isn't a pointer. It's really interesting how C++ handles these situations. I haven't encountered anything like it in other languages.

Reply:
Yep, that's exactly it. I think it's quite cool how C++ allows you to make these choice, as it's useful if performance is relevant.  And I think this is a function where we definitely don't want to pass any heavy objects (though, as I said, assuming optional is a pointer type, which I think it is, passing by value would be equally fine here).
I think in general, passing by value is a good default too, and if I remember correctly, it often allows better optimisations by the compiler. It can get quite messy too though - see e.g. https://twitter.com/herbsutter/status/1422644603883192320/photo/1. I think there's a couple of talks from Herb Sutter where he explains this, but I don't remember which one exactly.